File: tools/modal_spike/deno_probe.ts
// Headless probe: does the SlicerLive TS/WebGPU renderer run under Deno on a Modal GPU
// container, and how fast? Run inside the container by deno_probe.py; also runnable locally:
//   deno run --unstable-webgpu -A tools/modal_spike/deno_probe.ts /tmp/out
//
// Answers, in order: (1) is the adapter the real NVIDIA GPU (Vulkan) or a software fallback,
// (2) does a synthetic multi-field scene render, (3) how long does the REAL multi-volume
// scene (CTACardio + Panoramix + gizmo) take to stream in and trace at streaming sizes.
import { initDevice } from "../../render/device.ts";
import { SceneRenderer } from "../../render/scene-renderer.ts";
import { encodePNG } from "../../render/png.ts";
import { buildDualSphereFields } from "../../render/demos/multi-scene.ts";
import { buildMultiVolume } from "../../render/demos/selftest-scenes.ts";
import { TransformGizmoField } from "../../render/transform-gizmo-field.ts";
import type { Vec3 } from "../../render/mat4.ts";

const outDir = Deno.args[0] ?? "/tmp/out";
await Deno.mkdir(outDir, { recursive: true });
const R: Record<string, unknown> = {};
const ms = (t: number) => Math.round(performance.now() - t);

// ---- (1) adapter ----
let t = performance.now();
const gpu = await initDevice();
const info = (gpu.adapter as unknown as { info?: Record<string, string> }).info ?? {};
R.adapter = {
  vendor: info.vendor, architecture: info.architecture, device: info.device,
  description: info.description, backend: (info as { backend?: string }).backend,
  features: [...gpu.features],
};
R.initMs = ms(t);
console.log(`[probe] adapter ${JSON.stringify(R.adapter)} in ${R.initMs}ms`);

// ---- (2) synthetic multi-field scene ----
t = performance.now();
const syn = new SceneRenderer(gpu);
syn.build(buildDualSphereFields(gpu.device));
syn.setBackground(0.07, 0.08, 0.12);
syn.setCamera([30, -430, 150], [0, 0, 0], [0, 0, 1], 30, 640, 640);
const synRGBA = await syn.renderToRGBA(640, 640);
await Deno.writeFile(`${outDir}/synthetic.png`, await encodePNG(synRGBA, 640, 640));
let red = 0, blue = 0;
for (let i = 0; i < 640 * 640; i++) {
  const r = synRGBA[i * 4], b = synRGBA[i * 4 + 2];
  if (r > 90 && r > b + 20) red++;
  if (b > 90 && b > r + 20) blue++;
}
R.synthetic = { ms: ms(t), redPx: red, bluePx: blue };
console.log(`[probe] synthetic dual-sphere ${JSON.stringify(R.synthetic)}`);

// ---- (3) the REAL demo=multi scene: CTACardio + Panoramix + transform gizmo ----
t = performance.now();
let bytes = 0;
const sc = await buildMultiVolume(gpu.device, (n) => { bytes += n; });
R.load = { ms: ms(t), mb: +(bytes / 1e6).toFixed(1), mbps: +((bytes / 1e6) / ((performance.now() - t) / 1000)).toFixed(1) };
console.log(`[probe] scene stream ${JSON.stringify(R.load)}`);

t = performance.now();
const center: Vec3 = [
  (sc.cta.center[0] + sc.pano.center[0]) / 2,
  (sc.cta.center[1] + sc.pano.center[1]) / 2,
  (sc.cta.center[2] + sc.pano.center[2]) / 2,
];
const radius = Math.max(sc.cta.radius, sc.pano.radius) * 1.35;
const giz = new TransformGizmoField(sc.pano.field.worldCenter(), 88);
const scene = new SceneRenderer(gpu, "rgba8unorm");
scene.build([...sc.fields, giz]);
scene.setBackground(0.05, 0.06, 0.09);
R.buildMs = ms(t);
const dist = radius * 3;
const eye: Vec3 = [center[0], center[1] - dist, center[2]];

const median = (a: number[]) => a.slice().sort((x, y) => x - y)[a.length >> 1];
const sizes: [number, number][] = [[512, 512], [1024, 1024], [1920, 1080], [2560, 1440], [3840, 2160]];
const trace: Record<string, number> = {};
for (const [w, h] of sizes) {
  scene.setCamera(eye, center, [0, 0, 1], 30, w, h);
  const runs: number[] = [];
  for (let i = 0; i < 5; i++) { const t1 = performance.now(); await scene.traceSamples(w, h); runs.push(performance.now() - t1); }
  trace[`${w}x${h}`] = +median(runs).toFixed(1);
  console.log(`[probe] traceSamples ${w}x${h} median ${trace[`${w}x${h}`]}ms`);
}
R.traceMs = trace;

scene.setCamera(eye, center, [0, 0, 1], 30, 1024, 1024);
await Deno.writeFile(`${outDir}/multi.png`, await encodePNG(await scene.renderToRGBA(1024, 1024), 1024, 1024));

console.log("PROBE_JSON " + JSON.stringify(R));
gpu.device.destroy();

File: tools/modal_spike/deno_probe.py
"""
Does the SlicerLive renderer itself run on a Modal GPU container?

vulkan_probe.py already proved (2026-06-20, L4, driver 580.95.05) that Modal injects the NVIDIA
graphics libs and that -- once we write the Vulkan ICD / EGL vendor JSONs ourselves -- Vulkan
enumerates the real GPU. This probe reuses that image recipe and adds DENO, then runs the actual
TS/WebGPU code (render/, deno_probe.ts): adapter identity, a synthetic multi-field render, and the
REAL demo=multi scene (CTACardio + CTAAbdomenPanoramix + transform gizmo) -- stream-in time from
JetStream2 and traceSamples timings at streaming resolutions.

    modal run tools/modal_spike/deno_probe.py

PNGs land in tools/modal_spike/out/ for visual verification (they must match a local render).
"""

import json
import pathlib
import time

import modal

_IMPORT_TS = time.time()
DENO_VERSION = "v2.9.5"
# Local: the repo root, whose render/ tree we ship. In the container this module is re-imported
# from /root, where those parents don't exist — the sources are already at /app by then.
ROOT = pathlib.Path(__file__).resolve().parents[2] if modal.is_local() else pathlib.Path("/app")

app = modal.App("slicerlive-deno-probe")

image = (
    modal.Image.debian_slim(python_version="3.11")
    .apt_install(
        "libvulkan1", "vulkan-tools",          # Vulkan loader + vulkaninfo
        "mesa-utils-extra",                     # pulls the GL userspace libs NVIDIA needs
        "libglvnd0", "libx11-6", "libxext6",   # glvnd dispatch + Xlib (libGLX_nvidia links these)
        "curl", "unzip", "ca-certificates",
    )
    .run_commands(
        # Modal mounts the NVIDIA driver libs but NOT these vendor JSONs (vulkan_probe.py finding).
        "mkdir -p /usr/share/vulkan/icd.d /usr/share/glvnd/egl_vendor.d",
        "python3 -c \"import json;open('/usr/share/vulkan/icd.d/nvidia_icd.json','w')"
        ".write(json.dumps({'file_format_version':'1.0.0',"
        "'ICD':{'library_path':'libGLX_nvidia.so.0','api_version':'1.3.0'}}))\"",
        "python3 -c \"import json;open('/usr/share/glvnd/egl_vendor.d/10_nvidia.json','w')"
        ".write(json.dumps({'file_format_version':'1.0.0',"
        "'ICD':{'library_path':'libEGL_nvidia.so.0'}}))\"",
        # Deno: pinned binary, not the install script (deterministic image layer).
        f"curl -fsSL -o /tmp/deno.zip https://github.com/denoland/deno/releases/download/{DENO_VERSION}"
        "/deno-x86_64-unknown-linux-gnu.zip",
        "unzip -o /tmp/deno.zip -d /usr/local/bin && chmod +x /usr/local/bin/deno && rm /tmp/deno.zip",
    )
    .env({
        "NVIDIA_DRIVER_CAPABILITIES": "all",
        "NVIDIA_VISIBLE_DEVICES": "all",
        "DENO_DIR": "/tmp/deno",
        "DENO_NO_UPDATE_CHECK": "1",
    })
    # The renderer is dependency-free TS: the whole app is just these source trees
    # (render/ pulls in algorithms/ + logic/ through the scene builders).
    .add_local_dir(ROOT / "render", "/app/render")
    .add_local_dir(ROOT / "algorithms", "/app/algorithms")
    .add_local_dir(ROOT / "logic", "/app/logic")
    .add_local_file(ROOT / "tools/modal_spike/deno_probe.ts", "/app/tools/modal_spike/deno_probe.ts")
)


@app.function(image=image, gpu="L4", timeout=900)
def probe(launched_at: float) -> dict:
    import subprocess

    cold = time.time() - launched_at
    out = {"container_start_s": round(cold, 2)}

    def run(cmd, **kw):
        try:
            p = subprocess.run(cmd, capture_output=True, text=True, timeout=kw.get("timeout", 120))
            return (p.stdout + p.stderr).strip()
        except Exception as e:  # noqa: BLE001
            return f"(failed: {e})"

    out["nvidia_smi"] = run(["bash", "-lc", "nvidia-smi --query-gpu=name,driver_version --format=csv,noheader"])
    out["vulkaninfo"] = run(["bash", "-lc", "vulkaninfo --summary 2>/dev/null | grep -E 'deviceName|driverName|apiVersion' | head -6"])

    t0 = time.time()
    log = run(["deno", "run", "--unstable-webgpu", "-A", "/app/tools/modal_spike/deno_probe.ts", "/tmp/out"], timeout=800)
    out["deno_wall_s"] = round(time.time() - t0, 1)
    out["log"] = log
    for line in log.splitlines():
        if line.startswith("PROBE_JSON "):
            out["result"] = json.loads(line[len("PROBE_JSON "):])

    pngs = {}
    for name in ("synthetic.png", "multi.png"):
        p = pathlib.Path("/tmp/out") / name
        if p.exists():
            pngs[name] = p.read_bytes()
    out["pngs"] = pngs
    return out


@app.local_entrypoint()
def main():
    res = probe.remote(_IMPORT_TS)
    outdir = ROOT / "tools/modal_spike/out"
    outdir.mkdir(exist_ok=True)
    for name, data in res.pop("pngs", {}).items():
        (outdir / name).write_bytes(data)
        print(f"wrote {outdir / name} ({len(data)/1e3:.0f} kB)")
    print(res.pop("log", ""))
    print(json.dumps(res, indent=2))
